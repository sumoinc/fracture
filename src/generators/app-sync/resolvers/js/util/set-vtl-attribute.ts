import { ResourceAttributeGenerator } from "../../../../../services/resource-attribute";
import { StructureAttribute } from "../../../../../services/structure-attribute";
import { TypescriptStrategy } from "../../../../ts/strategy";

export const setVtlAttribute = (attribute: StructureAttribute) => {
  const strategy = TypescriptStrategy.of(attribute.project);
  const generator = attribute.generator;
  const name = strategy.formatAttributeName(attribute.name);
  const shortName = strategy.formatAttributeName(attribute.shortName);

  // non-generated
  if (generator === ResourceAttributeGenerator.NONE) {
    return `"${shortName}": $util.dynamodb.toDynamoDBJson($ctx.args.input.${name}),`;
    // generated values
  } else {
    switch (generator) {
      case ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP:
        return `"${shortName}": "$util.time.nowISO8601()"`;
      case ResourceAttributeGenerator.GUID:
        return `"${shortName}": "$util.autoId()"`;
      case ResourceAttributeGenerator.TYPE:
        return `"${shortName}": "unknown"`;

      default:
        throw new Error(`Unknown generator: ${generator}`);
    }
  }
};
