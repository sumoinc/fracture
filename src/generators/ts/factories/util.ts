import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
