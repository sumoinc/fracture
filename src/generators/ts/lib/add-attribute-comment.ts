import { Attribute } from "../../../model";
import { TypeScriptSource } from "../typescript-source";

export const addAttributeComment = (f: TypeScriptSource, a: Attribute) => {
  f.line(`/**`);
  a.comment.forEach((c) => f.line(` * ${c}`));
  f.line(` */`);
};
