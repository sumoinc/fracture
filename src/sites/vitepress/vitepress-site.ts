import { join } from "path";
import { SampleFile } from "projen";
import {
  TypeScriptProject,
  TypeScriptProjectOptions,
} from "projen/lib/typescript";
import { Site } from "../site";

export const filesToScaffold = [
  "api-examples.md",
  "index.md",
  "markdown-examples.md",
  ".vitepress/config.mts",
  ".vitepress/theme/index.ts",
  ".vitepress/theme/style.css",
  ".vitepress/theme/Layout.vue",
];

export class VitePressSite extends Site {
  constructor(parent: TypeScriptProject, options: TypeScriptProjectOptions) {
    super(parent, {
      outdir: options.name,
      ...options,
    });

    this.addDevDeps("vitepress");
    this.addGitIgnore(".vitepress/dist");
    this.addGitIgnore(".vitepress/cache");

    /**
     * !!! BIG CHEAT !!!
     *
     * These files are adapted from scaffold(), found here:
     * https://github.com/vuejs/vitepress/blob/main/src/node/init/init.ts
     *
     * Templates are all found here:
     * https://github.com/vuejs/vitepress/tree/main/template
     *
     **/
    filesToScaffold.map((file) => {
      new SampleFile(this, file, {
        sourcePath: join(__dirname, "template", file),
      });
    });

    /***
     * Add some shortcut commends that vitepress devs will expect to exist.
     */
    // add commands to run vitepress
    this.addTask("docs:dev", {
      exec: "pnpm vitepress dev",
    });
    this.addTask("docs:build", {
      exec: "pnpm vitepress build",
    });
    this.addTask("docs:preview", {
      exec: "pnpm vitepress preview",
    });

    // make sure we compile docs as part of standard default build
    // parent.compileTask.spawn(this.buildTask);

    /*
    parent.buildWorkflow?.addPostBuildSteps({
      name: `Build ${this.name}`,
      run: `npx projen build`,
      workingDirectory: `./${mergedOptions.outdir}`,
    });*/

    /**
     * Make sure we have a deployment workflow on the default branch to deploy
     * this site.
     */

    // const deployWorkflow =
    //   Pipeline.byBranchName(parent, mergedOptions.defaultReleaseBranch) ??
    //   new Pipeline(parent, { branchName: mergedOptions.defaultReleaseBranch });
    // deployWorkflow.paths.push(mergedOptions.outdir);

    // const vpArtifactName = `vp-${this.name}-artifact`;

    // // add vitepress output as an artifact
    // deployWorkflow.workflow.addPostBuildSteps({
    //   name: `Upload ${this.name} artifact`,
    //   uses: "actions/upload-artifact@v3",
    //   with: {
    //     name: vpArtifactName,
    //     path: join(mergedOptions.outdir, ".vitepress/dist"),
    //   },
    // });

    // deployWorkflow.workflow.addPostBuildJob(`deploy-${this.name}`, {
    //   runsOn: ["ubuntu-latest"],
    //   permissions: {
    //     contents: JobPermission.READ,
    //     idToken: JobPermission.WRITE,
    //   },
    //   steps: [
    //     {
    //       name: "Download build artifacts",
    //       uses: "actions/download-artifact@v3",
    //       with: {
    //         name: vpArtifactName,
    //         path: join(mergedOptions.outdir, ".vitepress/dist"),
    //       },
    //     },
    //     /*
    //       {
    //         name: "Configure AWS Credentials",
    //         uses: "aws-actions/configure-aws-credentials@v2",
    //         with: {
    //           "role-to-assume": `arn:aws:iam::${sdt.environment.accountNumber}:role/GitHubDeploymentOIDCRole`,
    //           "aws-region": sdt.environment.region,
    //         },
    //       },
    //       */
    //     {
    //       name: "Deploy",
    //       //run: `npx aws-cdk@${sdt.service.cdkDeps.cdkVersion} deploy --no-rollback --app ${sdt.service.cdkOutDistDir} ${sdt.stackPattern}`,
    //     },
    //   ],
    // });

    /*
    parent.release?.addJobs({
      bar: {
        needs: ["foo"],
        runsOn: ["ubuntu-latest"],
        permissions: {
          contents: JobPermission.READ,
          idToken: JobPermission.WRITE,
        },
        steps: [
          {
            name: "Configure AWS Credentials",
            uses: "aws-actions/configure-aws-credentials@v2",
            with: {
              "role-to-assume": `arn:aws:iam::${sdt.environment.accountNumber}:role/GitHubDeploymentOIDCRole`,
              "aws-region": sdt.environment.region,
            },
          },
          {
            name: "Deploy",
            //run: `npx aws-cdk@${sdt.service.cdkDeps.cdkVersion} deploy --no-rollback --app ${sdt.service.cdkOutDistDir} ${sdt.stackPattern}`,
          },
        ],
      },
    });
    */
  }
}
