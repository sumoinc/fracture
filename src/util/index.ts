export * from "./determineBuildId";
export * from "./test-util";
