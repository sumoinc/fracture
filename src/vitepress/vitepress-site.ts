import { join } from "path";
import { SampleFile } from "projen";
import {
  NodePackageManager,
  NodeProject,
  NodeProjectOptions,
} from "projen/lib/javascript";
import { SetOptional } from "type-fest";

export const filesToScaffold = [
  "api-examples.md",
  "index.md",
  "markdown-examples.md",
  ".vitepress/config.mts",
  ".vitepress/theme/index.ts",
  ".vitepress/theme/style.css",
  ".vitepress/theme/Layout.vue",
];

export class VitePressSite extends NodeProject {
  constructor(
    project: NodeProject,
    options: SetOptional<
      NodeProjectOptions,
      "name" | "defaultReleaseBranch"
    > = {}
  ) {
    super({
      name: "docs",
      outdir: "docs",
      defaultReleaseBranch: "main",
      license: options.license ?? "MIT",
      prettier: options.prettier ?? true,
      packageManager: options.packageManager ?? NodePackageManager.PNPM,
      pnpmVersion: options.pnpmVersion ?? "8",
      parent: project,
      ...options,
    });

    this.addDevDeps("vitepress");
    this.addGitIgnore(".vitepress/dist");
    this.addGitIgnore(".vitepress/cache");

    /**
     * !!! BIG CHEAT !!!
     *
     * These files are adapted from scaffold(), found here:
     * https://github.com/vuejs/vitepress/blob/main/src/node/init/init.ts
     *
     * Templates are all found here:
     * https://github.com/vuejs/vitepress/tree/main/template
     *
     **/

    filesToScaffold.map((file) => {
      new SampleFile(this, file, {
        sourcePath: join(__dirname, "template", file),
      });
    });

    // add commands to run vitepress
    this.addTask("docs:dev", {
      exec: "pnpm vitepress dev",
    });
    const build = this.addTask("docs:build", {
      exec: "pnpm vitepress build",
    });
    this.addTask("docs:preview", {
      exec: "pnpm vitepress preview",
    });

    this.defaultTask?.spawn(build);
  }
}
