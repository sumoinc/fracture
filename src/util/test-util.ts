import { Project } from "projen";
import { synthSnapshot } from "projen/lib/util/synth";

export const TEST_ACCOUNT_ONE = "000000000000";
export const TEST_ORG_ONE = "org-123456";
export const TEST_REGION_ONE = "us-east-1";

/**
 *
 * Helper to generate and return one file from fracture.
 * Useful when trying to test the final output of generated files.
 *
 * @param fracture
 * @param filepath
 * @returns collection of files starting with supplied filepath
 */
export const synthFiles = (project: Project, filepath: string = ""): any => {
  const snapshot = synthSnapshot(project);
  // console.log(Object.keys(snapshot));
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(filepath))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {} as { [key: string]: any });
  return filtered;
};

export const synthFile = (project: Project, filepath: string): string => {
  const files = synthFiles(project, filepath);

  if (files.length > 1) {
    throw new Error(`more than one file matched filepath "${filepath}"`);
  }

  if (files.length === 0) {
    throw new Error(`No files matched filepath "${filepath}"`);
  }

  return files[filepath];
};

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testProject = () => {
//   return new Project({
//     name: "my-project",
//   });
// };

// export const testNodeProject = (options: Partial<NodeProjectOptions> = {}) => {
//   return new NodeProject({
//     name: "my-project",
//     defaultReleaseBranch: "main",
//     packageManager: NodePackageManager.PNPM,
//     pnpmVersion: "8",
//     ...options,
//   });
// };

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testFractureProject = (
//   options: Partial<FractureProjectOptions> = {}
// ) => {
//   const fp = new FractureProject({
//     name: "my-project",
//     ...options,
//   });

//   /*
//   fp.buildEnvironment = new AwsEnvironment(fp, {
//     name: "test",
//     account: TEST_ACCOUNT_ONE,
//     region: TEST_REGION_ONE,
//     primaryHostedZoneId: "foo",
//     primaryDomainName: "bar",
//   });
//   */

//   return fp;
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testApp = () => {
//   return new App({
//     parent: testFractureProject(),
//     name: "my-app",
//   });
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testDataService = () => {
//   return new DataService(testApp(), {
//     name: "my-service",
//   });
// };

// /**
//  * Build a simple resource to test with
//  */
// export const testResource = () => {
//   const service = testDataService();
//   return new Resource(service.app, {
//     service,
//     name: "my-resource",
//   });
// };

// /**
//  * A more complicated resource configuration
//  */
// export const complexService = () => {
//   const service = testDataService();

//   const person = new Resource(service.app, { service, name: "person" });

//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     comments: ["A club"],
//     attributeOptions: [{ name: "name" }],
//   });

//   person.addAttribute({
//     name: "first-name",
//     shortName: "fn",
//   });
//   person.addAttribute({
//     name: "last-name",
//     shortName: "ln",
//   });
//   person.addAttribute({
//     name: "height",
//     shortName: "ht",
//     type: ResourceAttributeType.INT,
//   });
//   person.addAttribute({
//     name: "is-frog",
//     shortName: "if",
//     type: ResourceAttributeType.BOOLEAN,
//     comments: ["Is this person a frog?"],
//   });
//   person.addAttribute({
//     name: "fav-colors",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: ResourceAttributeType.STRING,
//     comments: ["List opf favorite colors"],
//   });

//   // one specific club
//   person.addAttribute({
//     name: "one-club",
//     type: club,
//   });
//   // an array of clubs
//   person.addAttribute({
//     name: "array-of-clubs",
//     shortName: "ad",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });

//   // an array of persons
//   club.addAttribute({
//     name: "array-of-persons",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });

//   // operation
//   Operation.create(person);
//   Operation.read(person);
//   Operation.update(person);
//   Operation.delete(person);

//   return service;
// };

// /**
//  * Build a simple resource attribute to test with
//  */
// export const testResourceAttribute = () => {
//   const service = testDataService();
//   const resource = testResource();
//   return new ResourceAttribute(service.app, {
//     name: "my-attr",
//     resource,
//   });
// };

// /**
//  * Build a simple structure to test with
//  */
// export const testStructure = () => {
//   const service = testDataService();
//   return new Structure(service.app, { name: "my-structure" });
// };

// /**
//  * Build a simple operation to test with
//  */
// export const testOperation = () => {
//   const resource = testResource();
//   return new Operation(resource.service.app, {
//     name: "my-operation",
//     resource,
//   });
// };

// /**
//  * Build a nuxt site to test with
//  */
// export const testNuxtJsSite = () => {
//   return new NuxtJsSite({
//     parent: testFractureProject(),
//     name: "my-site",
//   });
// };
// export const testServiceTwoRelatedResources = () => {
//   const service = testDataService();
//   const person = new Resource(service.app, {
//     service,
//     name: "person",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   club.addAttribute({
//     name: "people",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });
//   person.addAttribute({
//     name: "clubs",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });
//   return { service, person, club };
// };
