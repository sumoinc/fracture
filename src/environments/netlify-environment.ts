import { NodeProject } from "projen/lib/javascript";
import { ValueOf } from "type-fest";
import { Environment, EnvironmentOptions } from "./environment";
import { Settings } from "../core/fracture-settings";
import { AuthProvider, AuthProviderType } from "../workflows/auth-provider";

export interface NetlifyEnvironmentOptions extends EnvironmentOptions {
  /**
   * Github Secret name for the NETLIFY_AUTH_TOKEN
   *
   * @default NETLIFY_AUTH_TOKEN
   */
  readonly authTokenSecretName?: string;

  /**
   * Github Secret name for the NETLIFY_SITE_ID
   *
   * @default NETLIFY_SITE_ID
   */
  readonly siteIdSecretName?: string;
}

export class NetlifyEnvironment extends Environment {
  /**
   * Github Secret name for the NETLIFY_AUTH_TOKEN
   *
   * @default NETLIFY_AUTH_TOKEN
   */
  readonly authTokenSecretName: string;

  /**
   * Github Secret name for the NETLIFY_SITE_ID
   *
   * @default NETLIFY_SITE_ID
   */
  readonly siteIdSecretName: string;

  constructor(
    public readonly project: NodeProject,
    options: NetlifyEnvironmentOptions
  ) {
    super(project, options);

    // defaults
    this.authTokenSecretName =
      options.authTokenSecretName ?? "NETLIFY_AUTH_TOKEN";
    this.siteIdSecretName = options.siteIdSecretName ?? "NETLIFY_SITE_ID";
  }

  /*
  public get authProvider(): AuthProvider {
    return AuthProvider.fromNetlifyEnvironment(this.project, this);
  }
  */
}
