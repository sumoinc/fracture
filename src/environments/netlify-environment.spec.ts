import { NodeProject } from "projen/lib/javascript";
import { NetlifyEnvironment } from "./netlify-environment";
import { VitePressSite } from "../sites/vitepress/vitepress-site";
import { synthFile } from "../util/test-util";

describe("success conditions", () => {
  test("Smoke Test", () => {
    const project = new NodeProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });
    const environment = new NetlifyEnvironment(project, {
      name: "foo",
    });
    expect(environment).toBeTruthy();
  });

  test("with siteid", () => {
    const project = new NodeProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });
    const environment = new NetlifyEnvironment(project, {
      name: "foo",
      siteId: "bar",
    });
    expect(environment).toBeTruthy();

    // site options
    const vitePressSite = new VitePressSite(project, {
      name: "docs",
      defaultReleaseBranch: "main",
    });
    vitePressSite.deploy({
      branchPrefix: "feature",
      environment,
    });

    const content = synthFile(project, `.github/workflows/deployment.yml`);
    expect(content).toBeTruthy();
    expect(content).toMatchSnapshot();
    // console.log(content);
  });
});
