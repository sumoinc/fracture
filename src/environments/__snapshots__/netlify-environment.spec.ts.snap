// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success conditions with siteid 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: deploy
on:
  push:
    branches:
      - feature/*
jobs:
  build:
    name: Build
    runs-on: ubuntu-latest
    permissions:
      contents: write
    concurrency: build
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with: {}
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: latest
          cache: yarn
      - name: Install dependencies
        run: yarn install --check-files --frozen-lockfile
      - name: build
        run: npx projen build
      - name: Backup artifact permissions for \\"sites/docs/.vitepress/dist\\"
        run: getfacl -R . > permissions-backup.acl
        continue-on-error: true
        working-directory: sites/docs/.vitepress/dist
      - name: Upload artifacts for \\"sites/docs/.vitepress/dist\\"
        uses: actions/upload-artifact@v3
        with:
          name: build-artifact-7a1a4772
          path: sites/docs/.vitepress/dist
  deploy-docs-feature-foo:
    name: Deploy docs feature foo
    needs: build
    runs-on: ubuntu-latest
    permissions:
      contents: read
    concurrency: deploy-docs-feature-foo
    env:
      CI: \\"true\\"
    if: startsWith( github.ref, 'refs/heads/feature' )
    steps:
      - name: Download build artifacts for \\"sites/docs/.vitepress/dist\\"
        uses: actions/download-artifact@v3
        with:
          name: build-artifact-7a1a4772
          path: sites/docs/.vitepress/dist
      - name: Restore build artifact permissions for \\"sites/docs/.vitepress/dist\\"
        run: setfacl --restore=permissions-backup.acl
        continue-on-error: true
        working-directory: sites/docs/.vitepress/dist
      - name: Setup Node.js
        uses: actions/setup-node@v3
        with:
          node-version: latest
      - name: Deploy to Netlify
        uses: netlify/actions/cli@master
        env:
          NETLIFY_AUTH_TOKEN: \${{ secrets.NETLIFY_AUTH_TOKEN }}
          NETLIFY_SITE_ID: bar
        with:
          args: deploy --dir=sites/docs/.vitepress/dist --prod
"
`;
